src/app/InteractionModelEngine.h

```diff
@@ -0,0 +1,182 @@
+     * @param[in]  aEventCallback  	A function pointer for event call back
+     */
+    void SetEventCallback(void * const aAppState, const EventCallback aEventCallback);
+
+    /**
+     * @brief This is the default event handler to be called by application layer for any ignored or unrecognized event
+     *
+     * @param[in]  aEvent       A function pointer for event call back
+     * @param[in]  aInParam     A const reference to the input parameter for this event
+     * @param[out] aOutParam    A reference to the output parameter for this event
+     */
+    static void DefaultEventHandler(EventID aEvent, const InEventParam & aInParam, OutEventParam & aOutParam);
+
+    /**
+     * @brief Retrieve the singleton DataManagement Engine. Note this function should be implemented by the
+     *  adoption layer.
+     *
+     *  @return  A pointer to the shared InteractionModel Engine
+     *
+     */
+    static InteractionModelEngine * GetInstance(void);
+
+    InteractionModelEngine(void);
+
+    CHIP_ERROR Init(Messaging::ExchangeManager * apExchangeMgr);
+
+    void Shutdown();
+
+    CHIP_ERROR DeregisterClusterCommandHandler(chip::ClusterId aClusterId, chip::CommandId aCommandId,
+                                               Command::CommandRoleId aCommandRoleId);
+    CHIP_ERROR RegisterClusterCommandHandler(chip::ClusterId aClusterId, chip::CommandId aCommandId,
+                                             Command::CommandRoleId aCommandRoleId, CommandCbFunct aDispatcher);
+    void ProcessCommand(chip::ClusterId aClusterId, chip::CommandId aCommandId, chip::TLV::TLVReader & aReader,
+                        Command * apCommandObj, Command::CommandRoleId aCommandRoleId);
+
+    Messaging::ExchangeManager * GetExchangeManager(void) const { return mpExchangeMgr; };
+
+    CHIP_ERROR NewCommandSender(CommandSender ** const apComandSender);
+
+private:
+    void OnUnknownMsgType(Messaging::ExchangeContext * apEc, const PacketHeader & aPacketHeader, uint32_t aProtocolId,
+                          uint8_t aMsgType, System::PacketBufferHandle aPayload);
+    void OnInvokeCommandRequest(Messaging::ExchangeContext * apEc, const PacketHeader & aPacketHeader, uint32_t aProtocolId,
+                                uint8_t aMsgType, System::PacketBufferHandle aPayload);
+    void OnMessageReceived(Messaging::ExchangeContext * apEc, const PacketHeader & aPacketHeader, uint32_t aProtocolId,
+                           uint8_t aMsgType, System::PacketBufferHandle aPayload);
+    void OnResponseTimeout(Messaging::ExchangeContext * ec);
+
+    struct HandlerKey
+    {
+        HandlerKey(chip::ClusterId aClusterId, chip::CommandId aCommandId, Command::CommandRoleId aCommandRoleId);
+
+        chip::ClusterId mClusterId;
+        chip::CommandId mCommandId;
+        Command::CommandRoleId mCommandRoleId;
+        bool operator<(const HandlerKey & aOtherKey) const;
+    };
+
+    typedef std::map<HandlerKey, CommandCbFunct> HandlersMapType;
```

Comment:
Are we really OK to use std::map here? It's big and directly uses the heap, right? Believe we're planning to avoid the STL altogether.

Reply:
fully agree, I am planning to remove the map here in next PR after I figure out the corresponding ZAP change.

Reply:
https://github.com/project-chip/connectedhomeip/issues/4275